sqlglot/optimizer/merge_derived_tables.py

```diff
@@ -0,0 +1,246 @@
+from collections import defaultdict
+
+from sqlglot import expressions as exp
+from sqlglot.optimizer.scope import traverse_scope
+from sqlglot.optimizer.simplify import simplify
+
+
+def merge_derived_tables(expression):
+    """
+    Rewrite sqlglot AST to merge derived tables into the outer query.
+
+    Example:
+        >>> import sqlglot
+        >>> expression = sqlglot.parse_one("SELECT a FROM (SELECT x.a FROM x)")
+        >>> merge_derived_tables(expression).sql()
+        'SELECT x.a FROM x'
+
+    Inspired by https://dev.mysql.com/doc/refman/8.0/en/derived-table-optimization.html
+
+    Args:
+        expression (sqlglot.Expression): expression to optimize
+    Returns:
+        sqlglot.Expression: optimized expression
+    """
+    for outer_scope in traverse_scope(expression):
+        for subquery in outer_scope.derived_tables:
+            inner_select = subquery.unnest()
+            if (
+                isinstance(outer_scope.expression, exp.Select)
+                and isinstance(inner_select, exp.Select)
+                and _mergeable(inner_select)
+            ):
+                alias = subquery.alias_or_name
+                from_or_join = subquery.find_ancestor(exp.From, exp.Join)
+                inner_scope = outer_scope.sources[alias]
+
+                _rename_inner_sources(outer_scope, inner_scope, alias)
+                _merge_from(outer_scope, inner_scope, subquery)
+                _merge_joins(outer_scope, inner_scope, from_or_join)
+                _merge_expressions(outer_scope, inner_scope, alias)
+                _merge_where(outer_scope, inner_scope, from_or_join)
+                _merge_order(outer_scope, inner_scope)
+    return expression
+
+
+# If a derived table has these Select args, it can't be merged
+UNMERGABLE_ARGS = set(exp.Select.arg_types.keys()) - {
```

Comment:
keys is unnecessary, can just do set(arg_types)

Reply:
`.keys()` is in my muscle memory. I think it makes the intent more explicit.
I can remove though.

Reply:
it hurts performance, i've seen it happen in minerva 1.0